compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -174,6 +174,10 @@ impl<'a, 'ra, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'ra, 'tcx> {
                             );
                         }
                     }
+                    // HACK(mgca): see lower_const_item in ast_lowering
+                    ItemKind::Const(box ConstItem { body_id: Some(body_id), .. }) => {
+                        this.create_def(body_id, None, DefKind::AnonConst, i.span);
+                    }
```

Comment:
Shouldn't need to be changing def collector, it should just handle anon consts in the AST correctly already. Just need to change `ast::ConstItem` to store an `AnonConst` instead of NodeId/Expr as separate fields. This should also fix some of the ICEs you've been encountering